### DISCUSSION TITLE
feature: eglot-execute to optionally redirect executeCommand progress messages to a compilation-mode buffer

### DISCUSSION THREAD

**Discussion by adonovan** [joaotavora/eglot#1402] [General]
**Background:** Go's LSP server, gopls, replaces a number of tools that were formerly command-line tools such as gorename. An LSP server has many advantages: it has a hot in-memory cache, it can operate correctly on unsaved editor buffers, and it is of course tightly integrated into the editor. Rename has special support in the LSP. But there are a large variety of analysis tools we would like to build as features of gopls that need only the loosest of integration: at a minimum, the ability to stream text output to a terminal-like viewer, so that file:line:column line prefixes are marked up as source links, as in `compilation-mode`.

For example, Gopls today has a command, gopls.run_test, that can be invoked via LSP executeCommand. It runs `go test` on a particular test identified by source line, and sends back the mixed stdout/stderr stream via the LSP progress notification mechanism. Unfortunately, the command is not very useful to eglot users, neither out of the box nor after some elisp hacking, because of the way that eglot currently handles progress messages: it updates the echo area (or optionally the `*Messages*` buffer) after each notification. To be clear, this is not a criticism of eglot: this is a perfectly reasonable approach. But I think it would enable new categories of tool integration if there were a way to indicate that the output of certain commands should be handled differently: by streaming the output into a new, visible read-only buffer in compilation-mode, much the same way as the `compile` command streams the output of a subprocess. Initially I imagine that we would have to tell eglot which LSP commands to handle this way, but ideally I would like to make this a core feature of LSP, for example by adding a boolean to [executeCommandParams](https://microsoft.github.io/language-server-protocol/specifications/lsp/3.17/specification/#executeCommandParams) that indicates that the output should be displayed in a buffer. Other editors all have their own UI elements for streaming buffers; for example, VS Code has the Output panel with a selector that chooses among various streams.

**Proposal:** I propose that `eglot-execute` be changed so that certain commands are executed with a different progress notification handler that directs their output to a `compilation-mode` buffer. The set of commands would initially be configured by a customizable variable (i.e. client-side), but perhaps eventually it would be inferred from future metadata in the Command or CodeAction action argument (i.e. under the control of the server).

**Comment by joaotavora** [comment]
> but ideally I would like to make this a core feature of LSP, for example by adding a boolean to [executeCommandParams](https://microsoft.github.io/language-server-protocol/specifications/lsp/3.17/specification/#executeCommandParams) that indicates that the output should be displayed in a buffer.

Yes exactly, though "buffer" wouldn't be exactly the word you'd want to use.  Something dumber, more microsoftish or vscodish. In fact, you shouldn't even mention Emacs if you want to increase your chances, which I don't think are very high to begin with (I hope I'm wrong!)

>  The set of commands would initially be configured by a customizable variable (i.e. client-side),

No, let's use generic functions.  Can you show a transcript that *very clearly* (to save me time) of a LSP/JSONRPC session (real of fictional -- preferably real) that 1. shows an example of such command that wants to "output" to a buffer and 2. another command that doesn't want to output to a buffer. The best shape for the generic function can be decided then.  

Or maybe `eglot-handle-notification` is already the generic function that you want.  Keep in mind LSP/JSONRPC is a stateless protocol though.

**Reply by adonovan** [reply]
Understood, I will translate "buffer" into the appropriate LSP or VS Code term.

> Can you show a transcript...

Sure! Here's an example of a custom elisp command of mine that involves:
- the client calling `codeLens` in a `_test.go` file to enumerate the  set of runnable test functions in that file;
- the client then calling `executeCommand` to run one of them; then
- the server responding with a sequence of progress messages.

In this case, the server knows that any LSP Command with `command="gopls.test"` should be streamed to a buffer.

```
<-- textDocument/codeLens[4609] {"jsonrpc":"2.0","result":[{"range":{"start":{"line":13,"character":0},"end":{"line":13,"character":0}},"command":{"title":"run test","command":"gopls.test","arguments":["file:///Users/adonovan/w/xtools/gopls/internal/server/prompt_test.go",["TestAcquireFileLock"],null]}},{"range":{"start":{"line":53,"character":0},"end":{"line":53,"character":0}},"command":{"title":"run test","command":"gopls.test","arguments":["file:///Users/adonovan/w/xtools/gopls/internal/server/prompt_test.go",["TestReleaseAndAcquireFileLock"],null]}}],"id":4609}
--> workspace/executeCommand[4610] {"jsonrpc":"2.0","id":4610,"method":"workspace/executeCommand","params":{"title":"run test","command":"gopls.test","arguments":["file:///Users/adonovan/w/xtools/gopls/internal/server/prompt_test.go",["TestAcquireFileLock"],null]}}
  <-- window/workDoneProgress/create[13] {"jsonrpc":"2.0","method":"window/workDoneProgress/create","params":{"token":"1706860841872505163"},"id":13}
  --> window/workDoneProgress/create[13] {"jsonrpc":"2.0","id":13,"result":null}
  <-- $/progress {"jsonrpc":"2.0","method":"$/progress","params":{"token":"1706860841872505163","value":{"kind":"begin","title":"Running go test","cancellable":true,"message":"Running..."}}}
  <-- workspace/executeCommand[4610] {"jsonrpc":"2.0","result":null,"id":4610}
<-- $/progress {"jsonrpc":"2.0","method":"$/progress","params":{"token":"1706860841872505163","value":{"kind":"report","cancellable":true,"message":"=== RUN   TestAcquireFileLock"}}}
<-- window/logMessage {"jsonrpc":"2.0","method":"window/logMessage","params":{"type":3,"message":"2024/05/23 16:20:43 === RUN   TestAcquireFileLock\n\n\toperation=\"test\"\n"}}
<-- $/progress {"jsonrpc":"2.0","method":"$/progress","params":{"token":"1706860841872505163","value":{"kind":"report","cancellable":true,"message":"--- PASS: TestAcquireFileLock (0.00s)\nPASS"}}}
<-- window/logMessage {"jsonrpc":"2.0","method":"window/logMessage","params":{"type":3,"message":"2024/05/23 16:20:43 --- PASS: TestAcquireFileLock (0.00s)\nPASS\n\n\toperation=\"test\"\n"}}
<-- $/progress {"jsonrpc":"2.0","method":"$/progress","params":{"token":"1706860841872505163","value":{"kind":"report","cancellable":true,"message":"ok  \tgolang.org/x/tools/gopls/internal/server\t0.367s"}}}
<-- window/logMessage {"jsonrpc":"2.0","method":"window/logMessage","params":{"type":3,"message":"2024/05/23 16:20:43 ok  \tgolang.org/x/tools/gopls/internal/server\t0.367s\n\n\toperation=\"test\"\n"}}
<-- window/showMessage {"jsonrpc":"2.0","method":"window/showMessage","params":{"type":3,"message":"all tests passed"}}
<-- $/progress {"jsonrpc":"2.0","method":"$/progress","params":{"token":"1706860841872505163","value":{"kind":"end","message":"completed"}}}
```

The log reconstituted from the progress messages is this:
```
=== RUN   TestAcquireFileLock
--- PASS: TestAcquireFileLock (0.00s)
PASS
ok  	golang.org/x/tools/gopls/internal/server	0.284s
Info: all tests passed
```

And here's an example of the converse:

A codeAction request returned an action of kind source.doc with a Command of `command="gopls.doc"`. The server knows that such commands do not want streaming to a buffer. Invoking executeCommand causes the server to send a showDocument request, and no progress notifications at all.


```
 --> textDocument/codeAction[4708] {"jsonrpc":"2.0","id":4708,"method":"textDocument/codeAction","params":{"textDocument":{"uri":"file:///Users/adonovan/w/xtools/gopls/internal/server/code_action.go"},"range":{"start":{"line":297,"character":0},"end":{"line":297,"character":0}},"context":{"diagnostics":[],"only":["source.doc"]}}}
   <-- textDocument/codeAction[4708] {"jsonrpc":"2.0","result":[{"title":"View package documentation","kind":"source.doc","command":{"title":"View package documentation","command":"gopls.doc","arguments":[{"uri":"file:///Users/adonovan/w/xtools/gopls/internal/server/code_action.go","range":{"start":{"line":297,"character":0},"end":{"line":297,"character":0}}}]}}],"id":4708}
 --> workspace/executeCommand[4709] {"jsonrpc":"2.0","id":4709,"method":"workspace/executeCommand","params":{"title":"View package documentation","command":"gopls.doc","arguments":[{"uri":"file:///Users/adonovan/w/xtools/gopls/internal/server/code_action.go","range":{"start":{"line":297,"character":0},"end":{"line":297,"character":0}}}]}}
   <-- window/showDocument[14] {"jsonrpc":"2.0","method":"window/showDocument","params":{"uri":"http://127.0.0.1:60918/gopls/Gmbv8xFuqHE/pkg/golang.org/x/tools/gopls/internal/server?view=1","external":true,"takeFocus":true},"id":14}
   --> window/showDocument[14] {"jsonrpc":"2.0","id":14,"result":{"success":true}}
   <-- workspace/executeCommand[4709] {"jsonrpc":"2.0","result":null,"id":4709}
```


The way I would implement it is by having executeCommand observe the special flag, wait for the 'start' progress notification, then create a new buffer and save it in a map keyed by the progress token's id. Subsequent progress notifications would consult the map and append to the appropriate buffer; the final 'end' token would delete the map entry. If no 'start' progress token appears within 10s the executeCommand would be aborted.

**Reply by nemethf** [reply]
> > but ideally I would like to make this a core feature of LSP, for example by adding a boolean to [executeCommandParams](https://microsoft.github.io/language-server-protocol/specifications/lsp/3.17/specification/#executeCommandParams) that indicates that the output should be displayed in a buffer.

If we are talking about extending the language server protocol, 
Gopls developers might like to take a look at the approach of rust-analyzer:  
https://github.com/rust-lang/rust-analyzer/blob/master/docs/dev/lsp-extensions.md#runnables

I think `experimental/runnables` would work in this case.

**Reply by joaotavora** [reply]
> I think experimental/runnables would work in this case.

It does seem more promising. 

@adonovan, I don't think that (your) approach can work. Like I hinted,  `window/logMessage` doesn't have any concept of state so we cannot know what those messages refer to easily.  In that aspect `$/progress` is superior since it has a `token`/handle argument to address a previously created "progress reporter".  A similar mechanism should exist for creating and addressing a "logging outlet", i.e. an output stream.

But look at rust-analyzer's extensions, like @nemethf suggests. The more heavyweight servers throw their weight behind an extension, the more likely that extensions makes it into LSP, I'd say.

**Reply by joaotavora** [reply]
> It does seem more promising.

He wrote without even checking it out.  Seems that LSP upstream shut the idea down :-/ and I also don't see how it could solve the logging outlet problem.  But they probably had something figured for that, I guess, right?

> $/progress is superior 

This is what I'm talking about: https://microsoft.github.io/language-server-protocol/specifications/lsp/3.17/specification/#progress

**Reply by nemethf** [reply]
> > It does seem more promising.
> 
> He wrote without even checking it out. Seems that LSP upstream shut the idea down :-/ and I also don't see how it could solve the logging outlet problem. But they probably had something figured for that, I guess, right?

`go test` is probably similar to `cargo test`, so I think gopls could use `experimental/runnables`  here.  Just look at the screenshots of the eglot-x implementation [here](https://github.com/nemethf/eglot-x?tab=readme-ov-file#experimental-extensions) by searching for "runnables". 

Note: it is the client's responsibility to execute the command sent by the server.  There is also a [test explorer](https://github.com/rust-lang/rust-analyzer/blob/master/docs/dev/lsp-extensions.md#test-explorer) extension where the tests are executed by the server.

**Reply by joaotavora** [reply]
> Note: it is the client's responsibility to execute the command sent by the server.

Ah, then I think it's totally different from what @adonovan is proposing.  He wants the server to run the command and send output back to the client.  In Alan's model, as far as I can tell, it's the client is merely tasked with displaying the output suitably. This probably makes sense when the server can be "nearer" to the code (At least in Emacs/TRAMP scenarios it it).  This model seems interesting, the Rust model -- with all due respect :-) -- seems more useless.  Why would I ask the server interface to execute a command that I can ask my own local `cargo` tool to execute, just for the server to tell me exactly which `cargo` invocation I should run?  Seems needlessly round-trippy.

Anyway, that is besides the point. The bottom line is that this functionality needs a new out-of-protocol notification like `$gopls/stdout` that is probably _not_ `window/logMessage` (but could be, if you aded a new out-of-protocol field to _that_ message).  

The idea of tokens and creation/deletions of states used in LSP's  `$/progress` can be reused.  The only thing missing is a nice indirection between lists of buffers and the tokens/handles used by the new `$gopls/stdout` messages. That indirection could be, a weak hash table of sorts.  Of maybe some slot in a custom CLOS class for gopls.  The generic function mechanism of the `eglot-handle-notification` API should manage that.  So it seems this can all be done in some kind of `eglot-gopls-extra-stuff.el`.  Or `eglot-x.el`, more likely.  Not in `eglot.el`, because `eglot.el` is, by design, for LSP (in protocol) things.

**Reply by adonovan** [reply]
I agree that `logMessage` is not helpful; the `$/progress` notification mechanism, with its tokens, is what I was referring to. (Strictly, I should have omitted the "Info: all tests passed" line from the reconstituted log since that information comes from logMessage, not `$/progress`.)

The Rust situation is different, I think: it aims to return commands to the client which would then call execve. In my case, the server is expected to run the command. The command need not be a subprocess (like `go test`), it could be any server-side function that writes a stream of text followed by a success/failure status.

I don't think the notification mechanism you called `$gopls/stdout` need be different from `$/progress`, since progress is already capable of carrying chunks of arbitrary text, and it seems more complicated to define a parallel mechanism. But if it pleases our LSP overlords, it's fine by me.

>  The only thing missing is a nice indirection between lists of buffers and the tokens/handles used by the new $gopls/stdout messages.

Couldn't it just be a simple global map? If executeCommand knows to wait for the 'start' progress notification for some commands, then it has all the information it needs to deal with the buffer.  Cleanup (and map removal) would occur either when the corresponding 'end' notification arrives. Is your concern that the server may forget to send an 'end' message?

**Reply by joaotavora** [reply]
> Couldn't it just be a simple global map? 
>  Is your concern that the server may forget to send an 'end' message?

It could, of course.  Do you guarantee cleanup?  Weak-keyed hash table?  With what key?  I think the server slot is easier, but the map works too.

**Reply by joaotavora** [reply]
> I don't think the notification mechanism you called $gopls/stdout need be different from $/progress, since progress is already capable of carrying chunks of arbitrary text, and it seems more complicated to define a parallel mechanism. 

Sounds fishy to use `$/progress` for this... It's main and only use I know is for "Work Done Progress"  But you're right there's a separate slot for an "anything", problem is there'snot a very good way to dispatch on that.

 https://microsoft.github.io/language-server-protocol/specifications/lsp/3.17/specification/#progress

There should at least be a `kind` field of type String in the standard for me to even consider putting something in `eglot.el`. Until then I recommend using a custom notification and coding a `eglot-gopls.el` (or adding this `kind` yourself and then adding an `:around` just for your server to the method that handles `$/progress`, but that's slightly harder).

**Comment by ypytomets** [comment]
Related, test discovery:
- https://rust-analyzer.github.io/book/contributing/lsp-extensions.html#test-explorer
- https://github.com/microsoft/language-server-protocol/issues/944
- Metals implementation:
    - https://scalameta.org/metals/docs/integrations/debug-adapter-protocol/#via-explicit-main-or-test-commands
    - https://scalameta.org/metals/docs/editors/vscode/#via-test-explorer
    - https://scalameta.org/metals/docs/integrations/test-explorer/
